### DISCUSSION TITLE
[Bug] Missing cursor when using focus() in viewzone click event

### DISCUSSION THREAD

**Discussion by deunlee** [microsoft/monaco-editor#3624] [Q&A] [answered]
### Reproducible in vscode.dev or in VS Code Desktop?

- [X] Not reproducible in [vscode.dev](https://vscode.dev) or VS Code Desktop

### Reproducible in the monaco editor playground?

- [ ] Not reproducible in [the monaco editor playground](https://microsoft.github.io/monaco-editor/playground.html)

### Monaco Editor Playground Code

```typescript
var jsCode = [
	'"use strict";',
	'function Person(age) {',
	'	if (age) {',
	'		this.age = age;',
	'	}',
	'}',
	'Person.prototype.getAge = function () {',
	'	return this.age;',
	'};'
].join('\n');

var editor = monaco.editor.create(document.getElementById('container'), {
	value: jsCode,
	language: 'javascript',
});

editor.changeViewZones(function (changeAccessor) {
	var domNode = document.createElement('div');
	domNode.style.background = 'lightgreen';
	changeAccessor.addZone({
		afterLineNumber: 3,
		heightInLines: 3,
		domNode: domNode
	});
});

editor.onMouseDown(event => {
    if (event.target.type === monaco.editor.MouseTargetType.CONTENT_VIEW_ZONE) {
        editor.setSelection(new monaco.Selection(1,1,1,1));
        editor.focus();
    }
});

editor.setSelection(new monaco.Selection(7,1,7,1));
editor.focus();
```

### Reproduction Steps

The cursor is focused at the beginning of line 7 when the attached code is executed.
It's normal. But, using the same code that uses setSelection() and focus() in the click event doesn't work.
When clicking the view zone filled with green, the cursor should be moved to line 1, but the cursor does not blink on the screen.
I think it was moved internally. But it doesn't flicker on the screen. Also, when I type, the text doesn't enter into line 1.

**Comment by tamayika** [comment, marked as answer]
How about to delay callback?

```js
editor.onMouseDown(event => {
    if (event.target.type === monaco.editor.MouseTargetType.CONTENT_VIEW_ZONE) {
        setTimeout(() => {
            editor.setSelection(new monaco.Selection(1,1,1,1));
            editor.focus();
        }, 0);
    }
});
```

**Comment by deunlee** [comment]
@tamayika Thanks! It works.
btw, I don't know why it works when using delayed callback.

**Comment by tamayika** [comment]
Probably monaco's callback(change selection by mouse position) is fired after user defined callback.
Delayed callback is enqueued to JS runtime queue, so will be called after monaco's callback.
